fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/rules/AbstractDataAccessRule.java

```diff
@@ -501,23 +501,32 @@ private static List<RelationalExpression> createIntersectionAndCompensation(@Non
      * Private helper method that computes the ordering of the intersection using matches and the common primary key
      * of the data source.
      * @param partitionOrderings partition we would like to intersect
-     * @return a {@link PartialOrder}  of {@link BoundKeyPart} representing a common intersection ordering.
+     * @return a {@link PartialOrder}  of {@link KeyPart}s representing a common intersection ordering.
      */
     @SuppressWarnings("java:S1066")
     @Nonnull
-    private static PartialOrder<BoundKeyPart> intersectionOrdering(@Nonnull final List<List<BoundKeyPart>> partitionOrderings) {
+    private static PartialOrder<KeyPart> intersectionOrderingPartialOrder(@Nonnull final List<List<BoundKeyPart>> partitionOrderings) {
 
         final var orderingPartialOrders =
                 partitionOrderings.stream()
                         .map(boundKeyParts -> {
-                            final var orderingKeyParts =
+                            final var boundKeyPartsPartitions =
                                     boundKeyParts.stream()
+                                            .collect(Collectors.partitioningBy(boundOrderingKey -> boundOrderingKey.getComparisonRangeType() == ComparisonRange.Type.EQUALITY));
+
+                            final var equalityBoundKeyParts = boundKeyPartsPartitions.get(true).stream()
+                                    .map(BoundKeyPart::getKeyPart)
+                                    .collect(ImmutableSet.toImmutableSet());
+                            final var orderingKeyParts =
+                                    boundKeyPartsPartitions.get(false)
+                                            .stream()
                                             .filter(boundOrderingKey -> boundOrderingKey.getComparisonRangeType() != ComparisonRange.Type.EQUALITY)
+                                            .map(BoundKeyPart::getKeyPart)
                                             .collect(ImmutableList.toImmutableList());
 
-                            return PartialOrder.<BoundKeyPart>builder()
+                            return PartialOrder.<KeyPart>builder()
                                     .addListWithDependencies(orderingKeyParts)
```

Comment:
It looks like we have a linear chain of dependencies between non-equal key parts, is that right? if so, I wonder if we can somehow optimise the partial order merging algorithm called at line 525

Reply:
Resolved offline